core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerNotifiable.java

```diff
@@ -38,28 +44,44 @@ public VariableListenerNotifiable(VariableListener<Solution_, ?> variableListene
         }
     }
 
-    public <Entity_> VariableListener<Solution_, Entity_> getVariableListener() {
-        return (VariableListener<Solution_, Entity_>) variableListener;
+    void resetWorkingSolution() {
+        variableListener.resetWorkingSolution(scoreDirector);
     }
 
-    public Collection<VariableListenerNotification> getNotificationQueue() {
-        return notificationQueue;
+    void closeVariableListener() {
+        variableListener.close();
+    }
+
+    void addNotification(VariableListenerNotification notification) {
+        if (notificationQueue.add(notification)) {
+            notification.triggerBefore(variableListener, scoreDirector);
+        }
+    }
+
+    void triggerAllNotifications() {
+        int notifiedCount = 0;
+        for (VariableListenerNotification notification : notificationQueue) {
+            notification.triggerAfter(variableListener, scoreDirector);
+            notifiedCount++;
+        }
+        if (notifiedCount != notificationQueue.size()) {
+            throw new IllegalStateException("The variableListener (" + variableListener.getClass()
+                    + ") has been notified with notifiedCount (" + notifiedCount
+                    + ") but after being triggered, its notificationCount (" + notificationQueue.size()
+                    + ") is different.\n"
+                    + "Maybe that variableListener (" + variableListener.getClass()
+                    + ") changed an upstream shadow variable (which is illegal).");
+        }
```

Comment:
How could the queue have been changed without the iteration throwing a `ConcurrentModificationException`?
Are you saying that we're using concurrent queues?